client/src/components/Modal.tsx

```diff
@@ -0,0 +1,87 @@
+import { useRef, useEffect, createContext, useContext, ReactNode } from "react";
+import { createPortal } from "react-dom";
+import Button from "./Button";
+
+interface ModalContextProps {
+  onClose: () => void;
+}
+
+interface ModalProps {
+  children: ReactNode;
+  isOpen: boolean;
+  onClose: () => void;
+}
+
+interface ModalHeaderProps {
+  title: string;
+}
+
+interface ModalBodyProps {
+  children: ReactNode;
+}
+
+interface ModalFooterProps {
+  children: ReactNode;
+}
+
+function useOutsideClick(
+  ref: React.RefObject<HTMLDivElement>,
+  handler: () => void
+) {
+  useEffect(() => {
+    function handleClickOutside(event: MouseEvent) {
+      if (ref.current && !ref.current.contains(event.target as Node)) {
+        handler();
+      }
+    }
+    document.addEventListener("mousedown", handleClickOutside);
+    return () => {
+      document.removeEventListener("mousedown", handleClickOutside);
+    };
+  }, [ref, handler]);
+}
+
+const ModalContext = createContext<ModalContextProps | undefined>(undefined);
+
+export function Modal({ children, isOpen, onClose }: ModalProps) {
```

Comment:
It’s good practice to support keyboard users by adding the ability to close the modal with the Escape key. This can be done by adding a keydown listener:

```
useEffect(() => {
  function handleKeyDown(event: KeyboardEvent) {
    if (event.key === "Escape") {
      onClose();
    }
  }
  document.addEventListener("keydown", handleKeyDown);
  return () => {
    document.removeEventListener("keydown", handleKeyDown);
  };
}, [onClose]);
```